pkg/winperfcounters/watcher.go

```diff
@@ -204,8 +255,35 @@ func cleanupScrapedValues(vals []CounterValue) []CounterValue {
 	return vals
 }
 
-func removeItemAt(vals []CounterValue, idx int) []CounterValue {
+func removeItemAt[C CounterValue | RawCounterValue](vals []C, idx int) []C {
 	vals[idx] = vals[len(vals)-1]
-	vals[len(vals)-1] = CounterValue{}
+	var zeroValue C
+	vals[len(vals)-1] = zeroValue
 	return vals[:len(vals)-1]
 }
+
+func (pc *perfCounter) collectDataForScrape() (bool, error) {
+	if err := pc.query.CollectData(); err != nil {
+		var pdhErr *win_perf_counters.PdhError
+		if !errors.As(err, &pdhErr) || (pdhErr.ErrorCode != win_perf_counters.PDH_NO_DATA && pdhErr.ErrorCode != win_perf_counters.PDH_CALC_NEGATIVE_DENOMINATOR) {
+			return false, fmt.Errorf("failed to collect data for performance counter '%s': %w", pc.path, err)
+		}
+
+		if pdhErr.ErrorCode == win_perf_counters.PDH_NO_DATA {
+			// No data is available for the counter, so no error but also no data
+			return false, nil
+		}
+
+		if pdhErr.ErrorCode == win_perf_counters.PDH_CALC_NEGATIVE_DENOMINATOR {
+			// A counter rolled over, so the value is invalid
+			// See https://support.microfocus.com/kb/doc.php?id=7010545
+			// Wait one second and retry once
+			time.Sleep(time.Second)
+			if retryErr := pc.query.CollectData(); retryErr != nil {
+				return false, fmt.Errorf("failed retry for performance counter '%s': %w", pc.path, err)
+			}
```

Comment:
can this be expensive?

Reply:
Typically not, however, this one second sleep should be under the control of the caller. This specific piece of code was only moved, the sleep was already there, it is showing as new here since it was moved to a new private function to avoid code duplication. I will open an issue to change this in a separate PR.

Reply:
Issue to follow up https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39851